### PR TITLE
Warn if `dsymutil` returns an error code

### LINKED ISSUE #78770: rustc silently ignores failures in dsymutil
On macOS rustc runs dsymutil to generate debug info files. This invocation incorrectly ignores dsymutil's output and exit status. rustc thus succeeds even if dsymutil fails or crashes.

Example: in #78657, dsymutil crashed and left behind a corrupt .dSYM file. rustc nevertheless succeeded, thereby confusing the user after the debugger ignored that bad debug info.

### PR DESCRIPTION
**Pull request by jryans**
This checks the error code returned by `dsymutil` and warns if it failed. It
also provides the stdout and stderr logs from `dsymutil`, similar to the native
linker step.

I tried to think of ways to test this change, but so far I haven't found a good way, as you'd likely need to inject some nonsensical args into `dsymutil` to induce failure, which feels too artificial to me. Also, https://github.com/rust-lang/rust/issues/79361 suggests Rust is on the verge of disabling `dsymutil` by default, so perhaps it's okay for this change to be untested. In any case, I'm happy to add a test if someone sees a good approach.

Fixes https://github.com/rust-lang/rust/issues/78770

### CODE REVIEW

**Comment by alexcrichton** [comment]
From a technical perspective this seems fine (although it'd be nice to unify the error reporting with the linker  error reporting to ensure everything gets printed the same way). I am not comfortable approving this though since it has the possibility of being a regression for many projects. I don't know if there are any or how many projects are actually failing dsymutil but are succeeding to compile today. If this were to land there wouldn't be any stable recourse for those projects to get compiling again.

Basically I suspect there will need to be some analysis and/or investigation done to see what the impact of a change like this may be. I don't know what it would be myself, unfortunately. @lcnr would you be ok finding a different reviewer for that aspect?

**Comment by lcnr** [comment]
of course, I don't know who else is at least somewhat knowledgeable about `dsymutil`.

Maybe r? @nagisa 

> @lcnr would you be ok finding a different reviewer for that aspect?

In general, if I reassign an issue it's mostly a guess on who might be the most fitting for the given changes.
If that guess ends up being incorrect I am of course fine with looking for someone else.

**Comment by nagisa** [comment]
My position is similar to that of @alexcrichton's. We lack information on how many silent failures we have today. Ordinarily we'd run crater here, but crater builds on Linux, not macOS...

The only way I can see this landing is if the dsymutil failure printed a warning and did not terminate the rest of compilation.

**Comment by jryans** [comment]
Thanks everyone for your feedback! 😄 I agree it seems hard to gather evidence of how many silent failures are happening here today, so failing compilation does seem risky. I have changed the message to a warning and allowed compilation to continue, which I believe is still useful so that at least the failure is presented instead of hidden.

@nagisa, this should be ready for another look.
